packages/xsnap/test/xsnap-eof.test.js

```diff
@@ -0,0 +1,290 @@
+});
+
+async function spawnReflectiveWorker(handleCommand) {
+  const exitedPKit = makePromiseKit();
+  /** @type {XsnapChildProcess | undefined} */
+  let xsnapProcess;
+
+  /** @type {typeof import('child_process').spawn} */
+  const spawnSpy = (...args) => {
+    // @ts-expect-error overloaded signature
+    const cp = proc.spawn(...args);
+    xsnapProcess = cp;
+    return cp;
+  };
+
+  const io = { spawn: spawnSpy, os: os.type(), fs, tmpName };
+  const vat = await xsnap({
+    ...options(io),
+    handleCommand,
+    stdout: 'pipe',
+    stderr: 'pipe',
+  });
+  await vat.evaluate(
+    'function handleCommand(message) { issueCommand(new Uint8Array().buffer); };',
+  );
+
+  assert(xsnapProcess);
+
+  const toXsnap = xsnapProcess.stdio[3];
+  const fromXsnap = xsnapProcess.stdio[4];
+  const stderrP = text(xsnapProcess.stderr);
+  const stdoutP = text(xsnapProcess.stdout);
+
+  xsnapProcess.on('error', err => {
+    exitedPKit.reject(err);
+  });
+
+  xsnapProcess.on('exit', (code, signal) => {
+    exitedPKit.resolve({ code, signal });
+  });
+
+  return {
+    vat,
+    xsnapProcess,
+    toXsnap,
+    fromXsnap,
+    stdoutP,
+    stderrP,
+    exitedP: exitedPKit.promise,
+  };
+}
+
+async function issueCommandAndWait(worker, beforeWait, afterWait) {
+  let beforeWaitError;
+  let afterWaitError;
+  const { stdoutP, stderrP, exitedP } = worker;
+
+  try {
+    // this may error out since we are intentionally
+    // breaking communications with the worker
+    await beforeWait(worker); // vat.issueStringCommand('0');
+  } catch (err) {
+    beforeWaitError = err;
+  }
+
+  // wait for the worker to exit
+  const [strOut, strErr, { code: vatExitCode, signal: vatExitSignal }] =
+    await Promise.all([stdoutP, stderrP, exitedP]);
+
+  try {
+    // this may error out since vat has died.
+    await afterWait(worker); // vat.close();
+  } catch (err) {
+    afterWaitError = err;
+  }
+  return {
+    beforeWaitError,
+    afterWaitError,
+    strOut,
+    strErr,
+    vatExitCode,
+    vatExitSignal,
+  };
+}
+
+async function issueCommandZero(worker) {
+  await worker.vat.issueCommand('0');
+}
+
+async function closeVat(worker) {
+  await worker.vat.close();
+}
+
+async function verifyStdError(t, results, expectedStderr) {
+  const {
+    beforeWaitError,
+    afterWaitError,
+    strOut,
+    strErr,
+    vatExitCode,
+    vatExitSignal,
+  } = results;
+
+  t.not(beforeWaitError, undefined, 'issueCommand() must produce an error');
+  t.not(afterWaitError, undefined, 'vat.close() must produce an error');
+  t.is(vatExitSignal, null, 'vat exit signal must be null');
+  t.is(
+    vatExitCode,
+    ExitCode.E_IO_ERROR,
+    'vat exit code must indicate an IO error',
+  );
+  t.is(strOut, '', 'stdout must be empty');
+  t.is(strErr, expectedStderr, 'stderr must indicate expected error');
+}
+
+const testInterruption = test.macro(
+  async (t, onRequest, beforeWait, afterWait, verifyResults) => {
```

Comment:
The `test(…)` calls might also read better if these parameters were ordered by chronological use.
```suggestion
  async (t, beforeWait, onRequest, afterWait, verifyResults) => {
```
```js
test(
  'xsnap-worker complains while trying to READ an answer when pipes are closed',
  testInterruption,
  worker => worker.vat.issueCommand('0'),
  async function onRequest(worker, _message) {
    // the worker is blocked on read here, so closing "toXsnap" pipe
    // should cause an immediate read error on worker side.
    worker.toXsnap.end();
    worker.toXsnap.destroy();
    return new Uint8Array();
  },
  worker => worker.vat.close(),
  (t, results) => verifyStdError(
    t,
    results,
    'Got EOF on netstring read. Has parent died?\n',
  ),
);
```